### PR TITLE
Align offloaded parameters before merging

### PR DESCRIPTION
**Pull request by blbadger**
Fixes issue #868  to allow for merging and unloading using `auto` device alignment.   

For models of sufficient size relative to memory constraints, choosing `auto` device alignment causes `accelerate` to perform lazy offloaded parameter loading just before a forward call. Calling `merge_and_unload` on a model with offloaded parameters results in errors due to these offloaded parameters.

This PR addresses that issue by activating the appropriate `accelerate` device alignment hook methods for each module that is to be merged.  This loads offloaded parameters which are otherwise on the "meta' device (if the model does not fit on the execution device) to allow merging and then offloads the merged parameters to avoid out-of-memory errors.

The implementation is fairly straightforward with one exception: it seems that we must avoid calling`self._replace_module(parent, target_name, new_module, target)` as doing so leads to a memory leak as the `new_module` does not have the appropriate `AlignDevicesHook` hook.  Another approach would be to add this hook and proceed with replacement, but this may not be necessary for successful merging right now.

The main benefit compared to the existing approach of loading models using `cpu` is that this method should succeed even when cpu cannot load the model in its entirety, as each module is loaded and merged and unloaded sequentially.

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Interesting, thanks a lot for working on this long standing issues. It would be fantastic to have a solution for this.

To make the CI pass, could you please run `make style` on your code?

One thing that would be great is if we could have a test for this, I wonder if it can be written without too much effort. We could still use a small model and set a low memory limit to force offloading, then check if merging works. WDYT?

**Comment by blbadger** [comment]
Hi @BenjaminBossan my pleasure! Thanks for the reminder, I ran `make style` and it looks like `black` fixed up a file so the CI should go green.

That sounds great to me! Do you mind my asking what hardware the test would be run on (I mean GPU and CPU or CPU and Disk or all three)?  I will work on a custom `accelerate` device assignment that offloads parameters for a few small models (~100M parameters or so if that sounds good) so we can add a proper test of merging.

I have a quick question for that test: while developing this I have been testing on equality before and after merging using a (large) model's output layer activations, do you think this is sufficient for the official test? We could also test hidden layer activation equality if that is thought to be necessary.

**Comment by BenjaminBossan** [comment]
> Do you mind my asking what hardware the test would be run on (I mean GPU and CPU or CPU and Disk or all three)?

Ideally, we can run this test on the normal GitHub runners, so I would suggest to only use CPU and disk, setting the max memory in a way that some weights are on the former and some on the latter.

> I will work on a custom `accelerate` device assignment that offloads parameters for a few small models (~100M parameters or so if that sounds good) so we can add a proper test of merging.

Let's keep it simple at the start, so a single small model should suffice.

> I have been testing on equality before and after merging using a (large) model's output layer activations, do you think this is sufficient for the official test? We could also test hidden layer activation equality if that is thought to be necessary.

I believe output should be enough, except if we have a reason to suspect that intermediate activations differ, which I don't see.

**Comment by blbadger** [comment]
Looks like the small model test is somewhat trickier than anticipated: small models such as `gpt2` seem to be incapable of loading LoRA parameters to offloaded modules (the LoRA weights are offloaded and then not re-loaded later) which is not a problem for larger models. 

A way to work around that problem could be to make a truncated version of a larger model  (say the first few transformer blocks of Llama) and use that instead.  But there is also an issue when trying to offload any model to the disk: loading LoRA params onto modules offloaded on the disk does not result in successful re-loading when the forward hook is called.  

To summarize, testing merging and unloading using a cpu/disk device/offload combination for small models appears to be quite difficult due to incompatibilities with `accelerate`.  

I'd be happy to share or add tests for large models.  As long as LoRA parameters are loaded via 

```python
model = PeftModel.from_pretrained(model, lora_checkpoint)
```

rather than `get_peft_model` (to match the issue description) the merging and unloading is successful (this is how I wrote.  Let me know what you think @BenjaminBossan !

**Comment by BenjaminBossan** [comment]
Thanks for investigating this further.

Indeed, I think there are some issues left with offloaded weights. I'm not quite sure why this only affects small models and not larger ones. It would be great to share these tests here, we may want to add them as tests that are for now expected to fail but should succeed in the future.

Regarding the passing tests with larger models, feel free to add them. We will have to see if they can be executed on our runners.

**Comment by blbadger** [comment]
Sounds good to me! I have added the GPU/CPU offload test to the testing suite, although it may be best to keep separate as it has library dependencies that other tests do not.  Let me know if you want the test moved somewhere else and I would be happy to do so.

For your convenience a colab version of the test may be found [here](https://colab.research.google.com/drive/1xxN6uGeYv2Lj0FvUNKAWc6Af17pK_MWC?usp=sharing)

In case it is helpful, here is a small model test (as it would be added to the test suite) that will fail now for reference:
```python
    def _future_test_offloaded_merge(self):
        # will currently fail as offloaded gpt2 does not offload LoRA params properly
        device_map = {
            "transformer.wte": "cpu",
            "transformer.wpe": "cpu",
            "transformer.drop": "cpu",
            "transformer.ln_f": "cpu",
            "lm_head": "cpu",
        }
        for i in range(6):
            device_map[f"transformer.h.{i}"] = "cpu"
        for i in range(6, 12):
            device_map[f"transformer.h.{i}"] = "disk"
        model = self.transformers_class.from_pretrained("gpt2", device_map=device_map)
        config = LoraConfig(
            r=8,
            lora_alpha=32,
            target_modules=["c_proj"],  
            lora_dropout=0.05,
            bias="none",
            task_type="CAUSAL_LM",
        )
        model = get_peft_model(model, config)
        input_tokens = torch.tensor([[15137, 4776, 290, 3598, 812, 2084]])  # "Four score and seven years ago"
        pre_merge_olayer = model(input_tokens)[0]
        model = model.merge_and_unload()
        post_merge_olayer = model(input_tokens)[0]
        self.assertTrue(torch.all(pre_merge_olayer == post_merge_olayer))
```

As a final note, I checked the issue and it seems that there is a desire for both merging and then saving via `model.save_pretrained()`.  Implementing the saving of offloaded parameters will have to be added (to `transformers`) in the future, for example see [this issue](https://github.com/huggingface/transformers/issues/20072).

File: src/peft/tuners/lora/model.py [hunk 1]
```diff
@@ -381,12 +382,18 @@ def _unload_and_optionally_merge(self, merge=True, progressbar: bool = False, sa
             if getattr(self.model, "quantization_method", None) == "gptq":
                 raise ValueError("Cannot merge LORA layers when the model is gptq quantized")
 
-        key_list = [key for key, _ in self.model.named_modules() if "lora" not in key]
+        km_list = [[key, module] for key, module in self.model.named_modules() if "lora" not in key]
         desc = "Unloading " + ("and merging " if merge else "") + "model"
-        for key in tqdm(key_list, disable=not progressbar, desc=desc):
+        for pair in tqdm(km_list, disable=not progressbar, desc=desc):
+            key, module = pair[0], pair[1]
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/model.py:388]:
```suggestion
        for key, module in tqdm(km_list, disable=not progressbar, desc=desc):
```

File: src/peft/tuners/lora/model.py [hunk 2]
```diff
@@ -381,12 +382,18 @@ def _unload_and_optionally_merge(self, merge=True, progressbar: bool = False, sa
...
+            # re-load module params if offloaded to the meta device
+            if hasattr(module, "_hf_hook") and isinstance(module._hf_hook, AlignDevicesHook):
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/model.py:390]: One potential problem I see here: When going through `model.named_modules()`, we get all the modules in the graph, not only the leaf nodes. That means that we also get the top level module which contains all sub-modules. Is it correct to call the hook on it?

File: src/peft/tuners/lora/model.py [hunk 3]
```diff
@@ -431,12 +438,20 @@ def _unload_and_optionally_merge(self, merge=True, progressbar: bool = False, sa
                         new_module = torch.nn.Linear(target.in_features, target.out_features, bias=bias)
                 if merge:
                     target.merge(safe_merge=safe_merge)
-                self._replace_module(parent, target_name, new_module, target)
+
+                if hasattr(module, "_hf_hook") and isinstance(module._hf_hook, AlignDevicesHook):
+                    module._hf_hook.post_forward(module, torch.tensor([]))
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/model.py:443]: When I tested this with the bloomz model, this line was never reached.

File: src/peft/tuners/lora/model.py [hunk 4]
```diff
@@ -431,12 +438,20 @@ def _unload_and_optionally_merge(self, merge=True, progressbar: bool = False, sa
...
+                else:
+                    self._replace_module(parent, target_name, new_module, target)
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/model.py:445]: Is it correct to put this in the `else` clause? Should it not always be called?

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for working on a test and sorry for the late reply.

So the test you added would definitely be too large to work. However, I think it should be possible to swap a smaller model, e.g. `bigscience/bloomz-560m`. I don't know why `gpt2` doesn't work, but maybe it's not really the size but something else.

Here is something that should, in theory, do the trick:

```python
from transformers import AutoModelForCausalLM, AutoTokenizer, AutoConfig
from peft import LoraConfig, get_peft_model
import torch
from accelerate import infer_auto_device_map, init_empty_weights

def test():
    model_id = "bigscience/bloomz-560m"
    tokenizer = AutoTokenizer.from_pretrained(model_id)
    config = AutoConfig.from_pretrained(model_id)
    with init_empty_weights():
        model = AutoModelForCausalLM.from_config(config)
    model.tie_weights()

    # offload most transformer modules
    torch.manual_seed(0)
    device_map = infer_auto_device_map(model, max_memory={0: "2GIB", "cpu": "50GIB"})
    assert 0 in device_map.values()
    assert "cpu" in device_map.values()

    model = AutoModelForCausalLM.from_pretrained(model_id, device_map=device_map)
    assert len({p.device for p in model.parameters()}) == 2

    config = LoraConfig(init_lora_weights=False)
    model = get_peft_model(model, config)
    model.eval()

    input_tokens = tokenizer.encode("Four score and seven years ago", return_tensors="pt")
    model = model.merge_and_unload()
    model(input_tokens)  # fails: RuntimeError: Tensor on device meta is not on the expected device cuda:0!
```

This model would be small enough to run on our GPU runners and it loads on two devices. However, inference fails here, even with the changes from your PR.

**Comment by blbadger** [comment]
Thanks for the review and for looking into this @BenjaminBossan ! 

LoRA Bloom merging unfortunately does fail, but the failure is not specific to merging: the LoRA parameters are offloaded improperly and don't get re-onloaded, so attempting a forward pass will result in the error `Cannot copy out of meta tensor; no data!`.  As you noted, this is apparent as the model's modules do not have the `AlignDevicesHook` that they should after offloading.

This problem is more clear if we load the lora parameters from a checkpoint: attempting

`PeftModel.from_pretrained(model, lora_checkpoint)`

with bloom fails due to parameters being sent from the meta device (note that `get_peft_model` does not catch this).  For this reason I am writing the test using LoRA params loaded via `from_pretrained`, as only LoRA models that are offloaded properly should be able to use this method.

I think you are correct that it is not a size problem but rather an architectural problem. It seems that if model parameters are not sufficiently nested then the LoRA parameters attached to a module are not tied to the `AlignDevicesHook` and cannot be onloaded.

I will go ahead and add a test with a truncated version of a model that has sufficiently nested modules (llama for one) that is compatible with offloading LoRA parameters, while keeping the size under 500MB.

**Comment by BenjaminBossan** [comment]
> the LoRA parameters are offloaded improperly and don't get re-onloaded

I wonder if we would have to check the layer's hf hooks when adding a LoRA adapter and then attach to the same hooks to that adapter too.

> if model parameters are not sufficiently nested then the LoRA parameters attached to a module are not tied to the `AlignDevicesHook` and cannot be onloaded.

This still is very unclear to me, do you have an idea how nesting would influence the attachment of the correct hooks?

> I will go ahead and add a test with a truncated version of a model that has sufficiently nested modules (llama for one) that is compatible with offloading LoRA parameters, while keeping the size under 500MB.

That's great, thanks a lot.

**Comment by blbadger** [comment]
No problem! Thanks for your patience as well with the small model test. I have added it to the test suite and [here](https://colab.research.google.com/drive/1xxN6uGeYv2Lj0FvUNKAWc6Af17pK_MWC?usp=sharing#scrollTo=6Q2J5wOZ4AEL&uniqifier=1) is the colab version.

You are correct that the device alignment hooks must be attached to the LoRA parameters! The `PeftModel.from_pretrained()` method first removes all hooks from each (base model) module before re-attatching to the new Peft model's modules.  This appears to be why models without proper nesting cannot use `.from_pretrained()`, as the model must be capable of attaching and removing hooks to complete this function.  The same detach and re-attachment procedure could in principle be applied to models with LoRA adapters loaded via `get_peft_model`.

As to why small models do not attach hooks to LoRA parameters in modules properly, I am somewhat unclear on this as well.  I suspect this is a nesting issue only because the device alignment hooks are attached and removed recursively, but have not tested this idea so it may be quite wrong.  

Quick final note: I opened a PR in the `transformers` library for saving to offloaded parameters, so the second part of issue #868 should be covered now as well.

File: src/peft/tuners/lora/model.py (hunk 4 above)
Comment (blbadger) [reply src/peft/tuners/lora/model.py:444]: The `_replace_module` function unfortunately leads to a memory leak in this case, due to copied tensors of parameters (via `.to(device)`).  This manifests when, say, a model with 11GB of parameters on a 15GB GPU calls this function and experiences OOM.  As this function is mostly a device assignment utility, we may not need it if the device alignment hooks are present.

File: src/peft/tuners/lora/model.py (hunk 2 above)
Comment (blbadger) [reply src/peft/tuners/lora/model.py:389]: The approach here was just to mirror the way the device alignment hook are added to the model, such that every module with a hook has that hook called.  Calling `pre_forward` on a module twice should not affect device alignment as far as I know.  Some models may work well with calling the hook only on the leaf node modules, but I don't know if this is generally true.

File: src/peft/tuners/lora/model.py (hunk 3 above)
Comment (blbadger) [reply src/peft/tuners/lora/model.py:442]: Note: c.f. the discussion hook problems for smaller modules with LoRA parameters

File: tests/testing_common.py [hunk 5]
```diff
@@ -567,6 +569,34 @@ def _test_merge_layers(self, model_id, config_cls, config_kwargs):
         logits_merged_from_pretrained = model_from_pretrained(**dummy_input)[0]
         self.assertTrue(torch.allclose(logits_merged, logits_merged_from_pretrained, atol=atol, rtol=rtol))
 
+    def _test_offload_merge(self):
```
Comment (BenjaminBossan) [inline tests/testing_common.py:572]: Note that this function is not being called right now. It needs to be called from a top class that is a `unittest.TestCase`, like the other tests in this class. And since it requires a GPU, it should be moved to `test_gpu_examples.py`.

**Review by BenjaminBossan — COMMENTED** [review]
> I have added a test with a 400M parameter version of llama to the test suite and [here](https://colab.research.google.com/drive/1xxN6uGeYv2Lj0FvUNKAWc6Af17pK_MWC?usp=sharing#scrollTo=6Q2J5wOZ4AEL&uniqifier=1) is the colab version. We can go smaller on the model size if you wish too.

Thanks. I think that size should work fine.

I took your test and ran it on the #1106 PR and for me, the tests passed. I wonder if we can wait for that PR to be merged, as it seems that it makes the hook handling in this PR obsolete (as it preserves the original layer). WDYT?

Since I still don't fully understand the issue, I can't say for sure if that PR will fix the issue completely or just this specific test.

**Comment by blbadger** [comment]
> I took your test and ran it on the #1106 PR and for me, the tests passed. I wonder if we can wait for that PR to be merged, as it seems that it makes the hook handling in this PR obsolete (as it preserves the original layer). WDYT?

You know what, that should be a good solution to both merging offloaded modules as well as offloading LoRA parameters for less-nested models!  For merging, I think it is the new layer creation from old that requires the on-loading of the old parameters (in this PR) but if we don't need to create a new layer then on-loading should not be necessary.  The smaller models suffer from failing to onload LoRA parameters when the base model params are onloaded, but if they are the same layer then that should no longer be a problem too.

I took the liberty of applying the test in this PR to a few other models using the latest commit of #1106  and it seems to fix the merging and offloading issue nicely for all.  As suspected it also seems to eliminate the small model LoRA compatibility problems too:)

Do you think the offloaded model merge test from this PR is worth adding?

**Comment by BenjaminBossan** [comment]
Fantastic, thanks for testing, I'm glad that the refactor could resolve the issues, which confirms my conviction that the design in that PR is cleaner than what we have right now. It has been merged, so the test should now pass on PEFT main.

> Do you think the offloaded model merge test from this PR is worth keeping? If so I would be happy to move it to `test_gpu_examples.py` and add the call.

Yes, 100%. Feel free to modify this PR or create a new one with the test. Having it should ensure that we don't break this in the future. IIUC, we can now also use an off-the-shelf model to test it, right?

**Comment by blbadger** [comment]
Of course! My opinion is that it is a much better approach in general, very nice work.  I will modify this PR to include only the test with the fixes from your review. 

> Yes, 100%. Feel free to modify this PR or create a new one with the test. Having it should ensure that we don't break this in the future. IIUC, we can now also use an off-the-shelf model to test it, right?

That is correct and after trying the test with llama-7b, bloomz-560m, and gpt2-124m the latest commit from PR #1106 passes for all.  I will swap out the current model for gpt2-124m for speed, although we can probably go with an even smaller model if you wish.

File: tests/testing_common.py (hunk 5 above)
Comment (blbadger) [reply tests/testing_common.py:572]: Test has been moved and is called from a class that inherits from `unittest.TestCase`

**Comment by blbadger** [comment]
On second thought I will create a new PR for simplicity

**Comment by BenjaminBossan** [comment]
Thanks for the update. Could you please resolve the merge conflict? It should be enough to just take the code from PEFT main.

**Comment by blbadger** [comment]
> Thanks for the updates. I had some issues with your branch, probably because `model.py` keeps on falling behind of main. Maybe it would be easier to create a new PR with just the test? Otherwise, rebasing and squashing may also help.

Creating a new PR sounds good to me! I will add one and link this PR for reference.

That error is typical of the old problem of a lack of nested modules causing a LoRA parameter loading issue.   I think this is a version problem, as the test internally seems to use `peft==0.6.0.dev0` (although I am not sure why this would be) whereas running the test externally and pulling from the head, ie `peft==0.6.3.dev0` and running the test externally passes.  Installing the the latter peft version in the dev environment leads to the internal test passing as well.  

> Before merging, we probably also need to put the adapter on our own HF repo at https://huggingface.co/peft-internal-testing for security reasons.

Sounds good to me.  I am happy to upload (don't think I have the permissions right now) but if it is easier for you then you are most welcome to.

**Comment by BenjaminBossan** [comment]
Superseded by #1190
